### PR TITLE
Helper for constructing a simple cmap4

### PR DESCRIPTION
**Pull request by rsheeter**
Add an api to write a cmap from a high level definition. Builds on #272.

I intend the internals of `from_mappings` to change over time, but the entrypoint to remain stable.

### CODE REVIEW

**Comment by rsheeter** [comment]
Per f2f this morning I am going to migrate some of the code from https://github.com/googlefonts/fontmake-rs/pull/134 to here.

File: write-fonts/src/tables/cmap.rs [hunk 1]
```diff
@@ -74,9 +154,29 @@ mod tests {
             panic!("Expected a cmap4 in {encoding_record:?}");
         };
 
-        assert_eq!(start_code, to_vec(cmap4.start_code()));
-        assert_eq!(end_code, to_vec(cmap4.end_code()));
-        assert_eq!(id_delta, to_vec(cmap4.id_delta()));
-        assert_eq!(id_range_offsets, to_vec(cmap4.id_range_offsets()));
+        // The spec example says entry_selector 4 but the calculation it gives seems to yield 2 (?)
```
Comment (anthrotype) [inline write-fonts/src/tables/cmap.rs:157]: > The spec example says entry_selector 4 but the calculation it gives seems to yield 2 (?)

FWIW, also fontTools.ttLib.ttFont::getSearchRange agrees with you

File: write-fonts/src/tables/cmap.rs [hunk 2]
```diff
@@ -2,62 +2,142 @@
...
+        for (cp, gid) in mappings.into_iter() {
+            let cp = (cp as u32).try_into().unwrap();
+            let next_in_run = (
+                prev.0.checked_add(1).unwrap(),
+                prev.1.checked_add(1).unwrap(),
+            );
+            let current = (cp, gid.to_u16());
+            // Codepoint and gid need to be continuous
+            if current != next_in_run {
+                // Start a new run
```
Comment (anthrotype) [inline write-fonts/src/tables/cmap.rs:81]: fonttools's `_c_m_a_p` module contain a `splitRange` function that attempts 

> to split a range of character codes into subranges with consecutive glyph IDs in such a way that the cmap4 subtable can be stored "most" efficiently

https://github.com/fonttools/fonttools/blob/444a349ef4dd1bae2d43f6ea0e06d73fb6466b9a/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L783

I wonder if we should do something like that here as well

File: write-fonts/src/tables/cmap.rs [hunk 3]
```diff
@@ -2,62 +2,142 @@
...
+impl CmapSubtable {
+    fn create_format_4(
+        lang: u16,
+        end_code: Vec<u16>,
+        start_code: Vec<u16>,
+        id_deltas: Vec<i16>,
+    ) -> Self {
+        if HashSet::from([start_code.len(), end_code.len(), id_deltas.len()]).len() != 1 {
+            panic!("uneven parallel arrays, very bad. Very very bad.");
+        }
```
Comment (cmyr) [inline write-fonts/src/tables/cmap.rs:27]: I would probably just write this as,

```rust
assert!(end_code.len() == start_code.len() && end_code.len() == id_deltas.len(), "msg");
```

this doesn't really matter, but importing HashSet and doing an allocation here just seems unnecessary.

File: write-fonts/src/tables/cmap.rs [hunk 4]
```diff
@@ -2,62 +2,142 @@
...
+        // which is equal to floor(log2(segCount)))
+        let entry_selector = (seg_count as f32).log2().floor();
+
+        // Spec: Maximum power of 2 less than or equal to segCount, times 2
+        // ((2**floor(log2(segCount))) * 2, where “**” is an exponentiation operator)
+        let search_range = 2u16.pow(entry_selector as u32).checked_mul(2).unwrap();
+
+        // if 2^entry_selector*2 is a u16 then so is entry_selector
+        let entry_selector: u16 = entry_selector as u16;
+        let range_shift: u16 = seg_count * 2 - search_range;
```
Comment (cmyr) [inline write-fonts/src/tables/cmap.rs:40]:
```suggestion
        let entry_selector = entry_selector as u16;
        let range_shift = seg_count * 2 - search_range;
```

The type here is trivially inferred, so I don't see much advantage in declaring it explicitly?

**Review by cmyr — APPROVED** [review]
looks good!

File: write-fonts/src/tables/cmap.rs (hunk 2 above)
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:81]: I think we should but perhaps as a followon

File: write-fonts/src/tables/cmap.rs [hunk 5]
```diff
@@ -2,62 +2,142 @@
...
+
+        let id_range_offsets = id_deltas.iter().map(|_| 0).collect();
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:42]: Oh interesting, I haven't seen that before. I like it.

File: write-fonts/src/tables/cmap.rs (hunk 2 above)
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:81]: Deferred to https://github.com/googlefonts/fontations/issues/282

File: write-fonts/src/tables/cmap.rs [hunk 6]
```diff
@@ -2,62 +2,142 @@
...
+    /// Generates a [cmap](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap) that is expected to work in most modern environments.
+    ///
+    /// For the time being just emits [format 4](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-4-segment-mapping-to-delta-values)
+    /// so we can drive towards compiling working fonts. In time we may wish to additionally emit format 12 to support
+    /// novel codepoints.
+    pub fn from_mappings(mappings: impl IntoIterator<Item = (char, GlyphId)>) -> Cmap {
+        let mut end_code: Vec<u16> = Vec::new();
+        let mut start_code: Vec<u16> = Vec::new();
+        let mut id_deltas: Vec<i16> = Vec::new();
+        let mut id_range_offsets: Vec<u16> = Vec::new();
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply write-fonts/src/tables/cmap.rs:69]: Fixed. Also realized I no longer use id_range_offsets at all.

**Pull request merged**